onnxruntime/core/providers/cpu/math/einsum_utils/einsum_auxiliary_ops.h

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This module hosts implementations and thin wrappers over other onnx operator implementations
+// that will be called from within the Einsum operator implementation
+
+#pragma once
+
+#include "core/util/math.h"
+#include "core/providers/cpu/tensor/transpose.h"
+#include "core/providers/cpu/reduction/reduction_ops.h"
+
+#include <vector>
+
+namespace onnxruntime {
+
+namespace EinsumOp {
+
+// Holds device specific implementations of the following ops
+// This is because the CPU kernel needs to call the CPU implementation
+// and the CUDA kernel will call the CUDA implementation and share the rest of the code
+
+namespace DeviceHelpers {
+
+// Data copy op - Copies raw data from the source tensor's buffer to the destination tensor's buffer
+using DataCopy = std::function<Status(const Tensor& input, Tensor& output)>;
+
+// Transpose op - Transposes given input based on data in `permutation`
+using Transpose = std::function<Status(const std::vector<size_t>& permutation, const Tensor& input,
+                                       Tensor& output, const TensorShape* input_shape_override, void* cublas_handle)>;
+
+// MatMul op - Multiplies two inputs of shapes [num_batches, M, K] and [num_batches, K, N]
+template <typename T>
+using MatMul = std::function<Status(const T* input_1_data, const T* input_2_data, T* output_data,
+                                    size_t left_stride, size_t right_stride, size_t output_stride,
```

Comment:
nit: would be slightly more structured to take a gsl::span for the input and output data but not critical.

Reply:
I believe we don't need span here and a pointer to the raw buffer would do. We don't iterate over the raw buffer directly but we pas it in to the EP specific math helper for MatMul and by the time the execution flow reaches this line of code, we would have made all validations in EinsumOp::MatMul. Is there a specific reason you recommend usage of span here ?

Reply:
It's lightweight and passes along the info of how big the buffer is so that if anything needed to verify it is the correct size could do so easily. e.g. if you were debugging a memory corruption being able to see the size from a span would be helpful. Not essential but means the implementation of the function doesn't have to blindly trust the caller to have gotten the validation correct.